### ISSUE TITLE
2D MSAA breaks rendering if there are no 2D nodes to render

### ISSUE THREAD

**Issue by balassanne** [godotengine/godot#84280]
### Godot version

4.2 beta 4

### System information

Godot v4.2.beta4 - Windows 10.0.22621 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 4070 Ti (NVIDIA; 31.0.15.4584) - AMD Ryzen 9 5900X 12-Core Processor (24 Threads)

### Issue description

I am not sure what is going on to be honest, I tried to figure out for several hours hoping that I could open a more clear issue. But I managed to figure out it is a bug by running the same project in two different version of GoDot.

In the project I attached, make sure you do not have Steam open, run the project, click create a lobby then click start game.

![image](https://github.com/godotengine/godot/assets/73620153/028c81fd-60d9-43f7-bd91-8a677cd88e73)

The same project will have two different behaviors: 

* 4.2beta3: 
It will load the scene in around 2 seconds, then holding "Tab" will show the scoreboard, pressing console will open the console, pressing it again will close it, game is clearly active no matter the state of UI, or the input, you can orbit the camera around the player.

* 4.2beta4: It will simply freeze black no matter how long you wait, the multiplayer lobby UI is hidden and screen will stay black until you either hold "Tab" to open the scoreboard, or press "`" to open the console. If you press any of those inputs, the game scene will appear, but the it will render just as long as you hold 'Tab' or as long as the console should be active (between two console presses). The game seems to be running behind, that is why I used "render" here.

This behavior is very strange, there are no errors or warning displayed in output, and I think it should really be looked into. I feel that something was broken badly in this beta 4.

### Steps to reproduce

In the project I attached, make sure you do not have Steam open, run the project, click create a lobby then click start game.

![image](https://github.com/godotengine/godot/assets/73620153/028c81fd-60d9-43f7-bd91-8a677cd88e73)

The same project will have two different behaviors: 

* 4.2beta3 (CORRECT BEHAVIOR): 
It will load the scene in around 2 seconds, then holding "Tab" will show the scoreboard, pressing console will open the console, pressing it again will close it, game is clearly active no matter the state of UI, or the input, you can orbit the camera around the player.

* 4.2beta4 (WRONG BEHAVIOR): 
It will simply freeze black no matter how long you wait, the multiplayer lobby UI is hidden and screen will stay black until you either hold "Tab" to open the scoreboard, or press "`" to open the console. If you press any of those inputs, the game scene will appear, but it will render just as long as you hold 'Tab' or as long as the console should be active (between two console presses). besides this, it stays frozen.

### Minimal reproduction project

[Project.zip](https://github.com/godotengine/godot/files/13223077/Project.zip)

_Production edit: Confirmed to be a regression from https://github.com/godotengine/godot/pull/83649_

**Comment by kleonc** [comment]
> ### Minimal reproduction project
> 
> [Project.zip](https://github.com/godotengine/godot/files/13223077/Project.zip)

Can't test because of an error after running:
![Godot_v4 2-beta3_win64_w0ENXx1Xf6](https://github.com/godotengine/godot/assets/9283098/805bf4a8-73b4-4bd1-b115-78632b8a2757)

Please provide a runnable/testable MRP.

**Comment by AThousandShips** [comment]
Then this is steam specific if it only works with the steam plugin, so third party

**Comment by Lippanon** [comment]
I took a look at this. Firstly, this is definitely not a minimal reproduction project, but I understand the struggle. Running without steam I could reproduce your freeze problem in the main scene "Playground.tscn". It's not necessary to run the project with a server, simply trying to run this local scene is enough to see something's wrong, but I tested with creating LAN sv and loading into the scene as you indicated.
I began by seeing what project settings you were running and started to disable things to look for the problem, it lies with MSAA 2D. Simply disabling it will let you load into the scene:
![image](https://github.com/godotengine/godot/assets/115422160/3600d5ea-1d14-47fb-99b3-9dd37509084a)
Then the game will slow down heavily once the balls in the scene come in contact with each other. I'm sure there's already an issue open for this. To fix that part I replaced the collision shape of the DefaultBall.tscn with a SphereShape instead of the ConvexPolygon you were using (there's really no need to use that for a sphere-like shape, the simpler the better when it comes to physics).
With these 2 changes you should find it will run smoothly. Camera moves fine, the turrets shoot, the scoreboard shows and hides with no issues.
I suggest you start a fresh project and bring only the minimal things needed to reproduce the issue with MSAA 2D as that seems to be the main cause of your problem. 
TLDR: It looks like something broke with MSAA 2D.

**Comment by Lippanon** [comment]
As a followup, I hypothesized that when there are no 2d objects in a scene, having MSAA 2d on would cause the black-screen/freeze, and indeed that seems to be the case. 
I've created a MRP to demonstrate, a simple 3d scene with everything default except MSAA turned on to 4x: 
[msaa_2d_bug.zip](https://github.com/godotengine/godot/files/13238058/msaa_2d_bug.zip)

Running the scene results in a black screen (Windows 10, GTX 970, Vulkan). 
Adding a label to the scene fixes it, and the scene runs fine. 
Perhaps someone should change the issue's name if confirmed.

**Comment by Calinou** [comment]
I can confirm this with @Lippanon's MRP on 4.2 beta 6afd32098 (Linux, GeForce RTX 4090). Thanks for carrying out this detective work :slightly_smiling_face:

**Comment by balassanne** [comment]
@Lippanon thank you for doing such an investigation, for a problem that I couldn't even figure where to start searching. And thank you for actually reading those instructions.... I will improve the files in future issues, understood the struggle, now I know about the scene play too, I have 10 days in this engine, so please forgive me for the upload. :)

**Comment by BastiaanOlij** [comment]
@Calinou @Lippanon that actually makes perfect sense. We fixed that we're now properly rendering into the 2D MSAA buffers as you would expect if that has been turned on, but if there are no 2D elements to render it will skip the 2D rendering step and thus never resolve 2D MSAA (before this fix, if you rendered both 3D and 2D elements, as we were copying the 3D result into the wrong buffer, we had a black screen with 2D elements on top and lost the 3D render).

I'll see if I can fix that it always performs a resolve if 2D MSAA is turned on, but while this will fix the issue it is important to underline that this is less efficient then turning 2D MSAA off so 3D resolves directly into the final render buffer.

I don't think there is a safe way to detect during 3D rendering that there are no 2D elements and that it thus should ignore the 2D MSAA setting. We don't determine this until after 3D rendering finishes.

**Comment by akien-mga** [comment]
While it appears to be a regression technically, after discussion I'm moving it to the 4.3 milestone as it doesn't seem to be a critical issue for 4.2. There's a priori not much reason to use 2D MSAA when not rendering any 2D elements, so the solution here is to turn off 2D MSAA.

**Comment by BastiaanOlij** [comment]
Just to add to @akien-mga remark, I think the fixes that were done should resolve the issue reported in the original description. That is worth testing. The new issue @Lippanon reported however remains (i.e. if 2D MSAA is enabled but there are no 2D elements to render, black screen), we have an idea on how to fix this but won't be able to implement it in time. However enabling 2D MSAA when there are **no** 2D elements to render is an unlikely scenario and easily worked around by turning 2D MSAA off.

**Comment by Lippanon** [comment]
It is not uncommon for games to have a hotkey to enter "immersive" mode (example The Long Dark), or "screenshot" mode where 2d/UI elements become hidden. I don't think this scenario is unlikely at all. It seems unreasonable to expect the developer to turn of 2D MSAA or such things when hiding 2D elements.
I tested just now and rendering breaks with a simple toggle of a label's visibility.
Arrow right ("ui_right") shows the label, the scene renders. Arrow left ("ui_left") hides it, breaking rendering.

[msaa_2d_bug.zip](https://github.com/godotengine/godot/files/13365858/msaa_2d_bug.zip)

**Comment by BastiaanOlij** [comment]
@Lippanon thats why we are planning on fixing it, but for now its easy to work around. Just disable MSAA2D when you hide the UI, and reenable it when you show the UI. It's also more performant that way so even after this issue is fixed, I would still recommend people do so.

We can't solve that one, in code we don't find out there is nothing to render until we've already gone down the MSAA 2D path.

**Comment by BastiaanOlij** [comment]
@akien-mga I think I managed the fix as discussed yesterday with Reduz and Clay. See #84957.
Seeing timing and that the issue can be solved by (temporarily) disabling 2D MSAA, it may be better to cherry pick this fix for 4.2.1 so we have ample time to verify it works as advertised.
